// @gooddata/react-components/dist/interfaces/PivotTable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var lodash_1 = require("lodash");
var ColumnEventSourceType;
(function (ColumnEventSourceType) {
    ColumnEventSourceType["AUTOSIZE_COLUMNS"] = "autosizeColumns";
    ColumnEventSourceType["UI_DRAGGED"] = "uiColumnDragged";
    ColumnEventSourceType["FIT_GROW"] = "growToFit";
})(ColumnEventSourceType = exports.ColumnEventSourceType || (exports.ColumnEventSourceType = {}));
var UIClick;
(function (UIClick) {
    UIClick[UIClick["CLICK"] = 1] = "CLICK";
    UIClick[UIClick["DOUBLE_CLICK"] = 2] = "DOUBLE_CLICK";
})(UIClick = exports.UIClick || (exports.UIClick = {}));
function isAttributeColumnWidthItem(columnWidthItem) {
    return (!lodash_1.isEmpty(columnWidthItem) &&
        columnWidthItem.attributeColumnWidthItem !== undefined);
}
exports.isAttributeColumnWidthItem = isAttributeColumnWidthItem;
function isMeasureColumnWidthItem(columnWidthItem) {
    return (!lodash_1.isEmpty(columnWidthItem) &&
        columnWidthItem.measureColumnWidthItem !== undefined);
}
exports.isMeasureColumnWidthItem = isMeasureColumnWidthItem;
function isMeasureLocatorItem(locator) {
    return !lodash_1.isEmpty(locator) && locator.measureLocatorItem !== undefined;
}
exports.isMeasureLocatorItem = isMeasureLocatorItem;
