// @gooddata/react-components/dist/helpers/model/widthItems.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var AttributeColumnWidthItemBuilder = /** @class */ (function () {
    function AttributeColumnWidthItemBuilder(attributeIdentifier, width) {
        var _this = this;
        this.aggregation = function (aggregation) {
            _this.attributeColumnWidthItem.aggregation = aggregation;
            return _this;
        };
        this.attributeColumnWidthItem = {
            attributeIdentifier: attributeIdentifier,
            width: width,
        };
    }
    return AttributeColumnWidthItemBuilder;
}());
exports.AttributeColumnWidthItemBuilder = AttributeColumnWidthItemBuilder;
var MeasureColumnWidthItemBuilder = /** @class */ (function () {
    function MeasureColumnWidthItemBuilder(measureIdentifier, width) {
        var _this = this;
        this.attributeLocators = function () {
            var attributeLocators = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                attributeLocators[_i] = arguments[_i];
            }
            var _a;
            (_a = _this.measureColumnWidthItem.locators).splice.apply(_a, [-2,
                0].concat(attributeLocators.map(function (attributeLocatorItem) { return ({
                attributeLocatorItem: attributeLocatorItem,
            }); })));
            return _this;
        };
        this.measureColumnWidthItem = {
            width: width,
            locators: [
                {
                    measureLocatorItem: {
                        measureIdentifier: measureIdentifier,
                    },
                },
            ],
        };
    }
    return MeasureColumnWidthItemBuilder;
}());
exports.MeasureColumnWidthItemBuilder = MeasureColumnWidthItemBuilder;
exports.attributeColumnWidthItem = function (attributeIdentifier, width) {
    return new AttributeColumnWidthItemBuilder(attributeIdentifier, width);
};
exports.measureColumnWidthItem = function (measureIdentifier, width) {
    return new MeasureColumnWidthItemBuilder(measureIdentifier, width);
};
